drivers/flash/flash_stm32f7x.c

```diff
@@ -0,0 +1,201 @@
+#include "flash_stm32.h"
+
+#define STM32F7X_SECTOR_MASK		((u32_t) 0xFFFFFF07)
+
+bool flash_stm32_valid_range(struct device *dev, off_t offset, u32_t len,
+			     bool write)
+{
+	ARG_UNUSED(write);
+
+	return flash_stm32_range_exists(dev, offset, len);
+}
+
+static int write_byte(struct device *dev, off_t offset, u8_t val)
+{
+	struct stm32f7x_flash *regs = FLASH_STM32_REGS(dev);
+	u32_t tmp;
+	int rc;
+
+	/* if the control register is locked, do not fail silently */
+	if (regs->cr & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->cr &= ~CR_PSIZE_MASK;
+	regs->cr |= FLASH_PSIZE_BYTE;
+	regs->cr |= FLASH_CR_PG;
+
+	/* flush the register write */
+	tmp = regs->cr;
+
+	*((u8_t *) offset + CONFIG_FLASH_BASE_ADDRESS) = val;
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->cr &= (~FLASH_CR_PG);
+
+	return rc;
+}
+
+static int erase_sector(struct device *dev, u32_t sector)
+{
+	struct stm32f7x_flash *regs = FLASH_STM32_REGS(dev);
+	u32_t tmp;
+	int rc;
+
+	/* if the control register is locked, do not fail silently */
+	if (regs->cr & FLASH_CR_LOCK) {
+		return -EIO;
+	}
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	if (rc < 0) {
+		return rc;
+	}
+
+	regs->cr &= STM32F7X_SECTOR_MASK;
+	regs->cr |= FLASH_CR_SER | (sector << 3);
+	regs->cr |= FLASH_CR_STRT;
+
+	/* flush the register write */
+	tmp = regs->cr;
+
+	rc = flash_stm32_wait_flash_idle(dev);
+	regs->cr &= ~(FLASH_CR_SER | FLASH_CR_SNB);
+
+	return rc;
+}
+
+int flash_stm32_block_erase_loop(struct device *dev, unsigned int offset,
+				 unsigned int len)
+{
+	struct flash_pages_info info;
+	u32_t start_sector, end_sector;
+	u32_t i;
+	int rc = 0;
+
+	rc = flash_get_page_info_by_offs(dev, offset, &info);
+	if (rc) {
+		return rc;
+	}
+	start_sector = info.index;
+	rc = flash_get_page_info_by_offs(dev, offset + len - 1, &info);
+	if (rc) {
+		return rc;
+	}
+	end_sector = info.index;
+
+	for (i = start_sector; i <= end_sector; i++) {
+		rc = erase_sector(dev, i);
+		if (rc < 0) {
+			break;
+		}
+	}
+
+	return rc;
+}
+
+int flash_stm32_write_range(struct device *dev, unsigned int offset,
+			    const void *data, unsigned int len)
+{
+	int i, rc = 0;
+
+	for (i = 0; i < len; i++, offset++) {
+		rc = write_byte(dev, offset, ((const u8_t *) data)[i]);
+		if (rc < 0) {
+			return rc;
+		}
+	}
+
+	return rc;
+}
+
+/*
+ * Different SoC flash layouts are specified in across various
+ * reference manuals, but the flash layout for a given number of
+ * sectors is consistent across these manuals, with one "gotcha". The
+ * number of sectors is given by the HAL as FLASH_SECTOR_TOTAL.
+ *
+ * The only "gotcha" is that when there are 24 sectors, they are split
+ * across 2 "banks" of 12 sectors each, with another set of small
+ * sectors (16 KB) in the second bank occurring after the large ones
+ * (128 KB) in the first. We could consider supporting this as two
+ * devices to make the layout cleaner, but this will do for now.
+ */
```

Comment:
That is not true for the STM32F769 which can be configured in two different way, with either 12 sectors (4*32K, 128K, 7*256K) or 24 sectors (twice 4*16K, 64K, 7*128K).